sites/funmooc/src/backend/base/static/funmooc/js/xiti.js

```diff
@@ -0,0 +1,118 @@
+   * 
+   * @param number siteId 
+   * @param function callback 
+   */
+  function loadSmartTag(siteId, callback) {
+    var script = document.createElement('script');
+    script.type = "text/javascript";
+    script.async = true;
+    script.onload = callback;
+    script.onerror = function (error) { throw error };
+    script.src = "https://tag.aticdn.net/" + siteId + "/smarttag.js"
+    document.body.append(script)
+  }
+
+  function extractLevel2AndChapters(levels2) {
+    var chapters = location.pathname
+      .substr(4) // remove language slug
+      .split('/')
+      .filter(function (slug) { return slug.length > 0 });
+    var level2_slug = chapters.shift();
+
+    if (level2_slug.startsWith('cours') && chapters.length === 0) {
+      // We are on the search page.
+      level2_slug = 'search_engine';
+    }
+
+    var level2 = levels2.findIndex(function (level) {
+      // If slug is undefined, this is the homepage
+      if (level2_slug === undefined) return true;
+      return level.some(function (l) { return l.includes(level2_slug) });
+    });
+
+    if (level2 < 0) return null;
+
+    var data = {
+      level2: level2,
+      name: level2_slug,
+    }
+
+    chapters.forEach(function (chapter, index) {
+      parameters["chapter" + index] = chapter;
+    });
+
+    return data;
+  }
+
+  function dispatchPageTag(tag, data) {
+    // Detail language
+    var lang = (document.documentElement.lang || '-').split('-')[0].toUpperCase();
+    if (lang) {
+      tag.customVars.set({ site: { 1: '[' + lang + ']' } });
+    }
+
+    // Detail search if there is
+    var search_query = new URL(location.href).searchParams.get('query');
+    if (search_query) {
+      tag.internalSearch.set({
+        keyword: search_query,
+      });
+    }
+
+    // Prepare page tag before dispatch
+    tag.page.set(data)
+
+    // Dispatch data
+    tag.dispatch();
+  }
+
+  /**
+   * spyClickOnButtons
+   * spy all button clicks on the current page
+   */
+  function spyClickOnButtons(tag, data) {
+    var tag = this;
+    document.getElementsByTagName('button').forEach(function ($button) {
+      $button.addEventListener('click', function (event) {
+        var $target = event.currentTarget;
+        var name = $target.innerText || $target.classList.value.replace(' ', '.');
+
+        var parameters = data;
+        parameters.elem = $target;
+        parameters.name = name;
+        parameters.type = 'action';
+
+        tag.click.send(parameters);
+      }, false)
+    });
+  }
+
+  /**
+   * initializeSmartTag
+   * 
+  */
+  function initializeSmartTag(siteId, levels2) {
+    if (ATInternet) {
+      var tag = new ATInternet.Tracker.Tag({ site: siteId, secure: true });
+      var tagData = extractLevel2AndChapters(levels2);
+
+      if (tagData) {
+        dispatchPageTag(tag, tagData)
+        // Ensure that react components are rendered...
+        setTimeout(spyClickOnButtons.bind(null, tag, tagData), 100);
```

Comment:
there's no event you can listen for this?

Reply:
There is something to do with `MutationObserver`.